### PR TITLE
HDDS-7231. Integrate the GetKeyInfo API to key read flows

### PR DESCRIPTION
**Pull request by duongkame**
## What changes were proposed in this pull request?
This is a part of the container location cache implementation(HDDS-7223), specified in [Client Interaction](https://docs.google.com/document/d/1-hVdA7mRBFiFOqr0J0Pg6-Ie5SOZaT7Z/edit#heading=h.4d34og8) design.

This PR includes the changes to integrate the GetKeyInfo API to key read flows, keeping client code backward compatible with previous OM versions.

https://issues.apache.org/jira/browse/HDDS-7231

## Details about client interaction changes:
**Before:**

1. Client calls lookupKey.
2. Client calls getBlock and readChunk. They when there's an exception in either calls:
 	- If it's a security problems, fail without retry.
	- If StorageContainerException (threw when Datanode returns any stattus code than), recall lookupKey to get latest block location info.
	- If it a conectivity issues (all datanodes are not reachable), or any IOException, just retry without recalling lookupKey.

**After:**

Client vs new OM.
1. Client calls getKeyInfo with `forceCacheUpdate = false` to tell OM to take container location from its cache.
2. Client calls getBlock and readChunk. When there's an exception in either call:
 	- If it's a security problems, fail without retry.
	- If StorageContainerException (threw when Datanode returns any stattus code than SUCCESS) or a conectivity issue (all datanodes are not reachable), recall getKeyInfo with `forceCacheUpdate = true` refresh block location from OM and tell OM to call SCM for updated container locations.
	- If any other IOException, just retry without recalling getKeyInfo.

Client vs old OM.
1. Client calls lookupKey.
2. Client calls getBlock and readChunk. When there's an exception in either calls:
 	- If it's a security problems, fail without retrying.
	- If StorageContainerException (threw when Datanode returns any stattus code than SUCCESS) or a conectivity issue (all datanodes are not reachable), recall lookupKey to get latest block location info.
	- If any other IOException, just retries without recalling lookupKey.

## How was this patch tested?
Standard CI: https://github.com/duongkame/ozone/actions/runs/3191413426/jobs/5207698694
- Integration tests are added to verify clients correctly 

Manual verification for the following cases:
- New client vs old OM.
- Old client vs new OM.

### CODE REVIEW

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java [hunk 1]
```diff
@@ -302,20 +310,24 @@ protected synchronized int readWithStrategy(ByteReaderStrategy strategy)
       try {
         numBytesRead = strategy.readFromBlock(current, numBytesToRead);
         retries = 0; // reset retries after successful read
+      } catch (SCMSecurityException ex) {
```
Comment (kerneltime) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java:313]: Whats the context for this change?

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java [hunk 2]
```diff
@@ -302,20 +310,24 @@ protected synchronized int readWithStrategy(ByteReaderStrategy strategy)
...
           throw e;
         }
-      } catch (SCMSecurityException ex) {
-        throw ex;
       } catch (IOException ex) {
         // We got a IOException which might be due
         // to DN down or connectivity issue.
         if (shouldRetryRead(ex)) {
-          current.releaseClient();
+          if (isConnectivityIssue(ex)) {
```
Comment (kerneltime) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java:326]: Can you add a few lines for the logic in place here.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java (hunk 1 above)
Comment (duongkame) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java:313]: It was there. I just move that piece up to separate it from other exceptions that we retry on.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java (hunk 2 above)
Comment (duongkame) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java:332]: Added comments along with the updated PR description.

**Comment by GeorgeJahad** [comment]
There is a call to looupKey here:
https://github.com/apache/ozone/blob/1ea4c2adc640a02508696bf346c7f1f7b6e0afd9/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java#L1264

Should that be changed to getKeyInfo()?

**Comment by GeorgeJahad** [comment]
There is a call to lookupFile() here:
https://github.com/apache/ozone/blob/1ea4c2adc640a02508696bf346c7f1f7b6e0afd9/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java#L1683
But lookupFile() doesn't appear to use the cache.  Should it?

File: hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java [hunk 3]
```diff
@@ -1437,6 +1421,34 @@ public OzoneKeyDetails getKeyDetails(
...
+        .setLatestVersionLocation(getLatestVersionLocation)
+        .setForceUpdateContainerCacheFromSCM(forceUpdateContainerCache)
+        .build();
+    return getKeyInfo(keyArgs);
+  }
+
+  private OmKeyInfo getKeyInfo(OmKeyArgs keyArgs) throws IOException {
+    final OmKeyInfo keyInfo;
+    if (omVersion.compareTo(OzoneManagerVersion.OPTIMIZED_GET_KEY_INFO) >= 0) {
+      keyInfo = ozoneManagerClient.getKeyInfo(keyArgs, false)
```
Comment (GeorgeJahad) [inline hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:1444]: when does the assumeS3Context parameter get set to true?

Comment (duongkame) [reply hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java:1444]: Great catch. The parameter will be used as per HDDS-7324 to improve S3G GET performance.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java [hunk 4]
```diff
@@ -0,0 +1,674 @@
...
+ * This class includes the integration test-cases to verify the integration
+ * between client and OM to keep container location cache eventually
+ * consistent. For example, when clients facing particular errors reading data
+ * from datanodes, they should inform OM to refresh location cache and OM
+ * should in turn contact SCM to get the updated container location.
+ *
+ * This integration verifies clients and OM using mocked Datanode and SCM
+ * protocols.
+ */
+public class TestOmContainerLocationCache {
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java:124]: I think this is a very nicely written class of tests.  A few minor nits below.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java [hunk 5]
```diff
@@ -0,0 +1,674 @@
...
+
+  /**
+   * Verify that in a happy case, container location is cached and reused
+   * in OM.
+   */
+  @Test
+  public void containerCachedInHappyCase() throws Exception {
+    byte[] data = "Test content".getBytes(UTF_8);
+
+    mockScmAllocation(testContainerId, 1L);
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java:254]: This method name could be clearer.  Instead of mockScmAllocation(), how about mockAllocateBlockOnDN1()  ?

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java [hunk 6]
```diff
@@ -0,0 +1,674 @@
...
+    mockScmGetContainerPipeline(testContainerId, DN1);
+
+    // Read keyName1.
+    OzoneKeyDetails key1 = bucket.getKey(keyName1);
+    verify(mockScmContainerClient, times(1))
+        .getContainerWithPipelineBatch(newHashSet(testContainerId));
+
+    try (InputStream is = key1.getContent()) {
+      mockGetBlock(mockDn1Protocol, testContainerId, 1L, data, null, null);
+      mockReadChunk(mockDn1Protocol, testContainerId, 1L, data, null, null);
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java:276]: these two mocks work for both key1 and key2 don't they?

if so, i would move them above the try because they look key1 as is.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java [hunk 7]
```diff
@@ -0,0 +1,674 @@
...
+      byte[] read = new byte[(int) key1.getDataSize()];
+      IOUtils.read(is, read);
+      Assertions.assertArrayEquals(data, read);
+    }
+
+    // Create keyName2.
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java:282]: maybe
```
// Create keyName2 in the same container so as to reuse the cache
```
Just to make it super clear what is happening here.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java [hunk 8]
```diff
@@ -0,0 +1,674 @@
...
+                            byte[] data,
+                            Exception exception,
+                            Result errorCode) throws Exception {
+
+    final CompletableFuture<ContainerCommandResponseProto> response;
+    if (exception != null) {
+      response = new CompletableFuture<>();
+      response.completeExceptionally(exception);
+    } else if (errorCode != null) {
+      ContainerCommandResponseProto getBlockReps =
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java:591]: getBlockResp ??

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java [hunk 9]
```diff
@@ -0,0 +1,674 @@
...
+                             long containerId, long localId,
+                             byte[] data,
+                             Exception exception,
+                             Result errorCode) throws Exception {
+    final CompletableFuture<ContainerCommandResponseProto> response;
+    if (exception != null) {
+      response = new CompletableFuture<>();
+      response.completeExceptionally(exception);
+    } else if (errorCode != null) {
+      ContainerCommandResponseProto readChunkReps =
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java:638]: readChunkResp??

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/MiniOzoneClusterImpl.java [hunk 10]
```diff
@@ -904,14 +904,14 @@ private void configureSCMheartbeat() {
     }
 
     private void configureOM() {
-      conf.set(OMConfigKeys.OZONE_OM_ADDRESS_KEY, "127.0.0.1:0");
-      conf.set(OMConfigKeys.OZONE_OM_HTTP_ADDRESS_KEY, "127.0.0.1:0");
+      conf.set(OMConfigKeys.OZONE_OM_ADDRESS_KEY, "127.0.0.1:12349");
+      conf.set(OMConfigKeys.OZONE_OM_HTTP_ADDRESS_KEY, "127.0.0.1:12344");
```
Comment (GeorgeJahad) [inline hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/MiniOzoneClusterImpl.java:908]: why were these changed?

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockInputStream.java [hunk 11]
```diff
@@ -68,7 +72,6 @@
 /**
  * Tests for {@link BlockInputStream}'s functionality.
  */
-@RunWith(MockitoJUnitRunner.class)
```
Comment (GeorgeJahad) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockInputStream.java:71]: Just curious, why was this removed?

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockInputStream.java [hunk 12]
```diff
@@ -280,35 +284,9 @@ public void testRefreshPipelineFunction() throws Exception {
...
-            false, null, chunks, chunkDataMap, isRefreshed,
-            new IOException("unavailable"));
-    try {
-      Assert.assertFalse(isRefreshed.get());
-      byte[] b = new byte[200];
-      blockInputStreamWithRetry.read(b, 0, 200);
-      // As in case of IOException we do not do do refresh.
-      Assert.assertFalse(isRefreshed.get());
-      Assert.assertTrue(logCapturer.getOutput().contains(
-          "Retry to get chunk info fail"));
```
Comment (GeorgeJahad) [inline hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockInputStream.java:304]: This is the only test that excercises the log, and it's been removed.  Is that a problem?

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockInputStream.java (hunk 11 above)
Comment (duongkame) [reply hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockInputStream.java:71]: The `MockitoJUnitRunner` doesn't work with `ParameterizedTest`. So, I removed that declaration to let JUnit figure out the right runner itself. A little side effect is that we have to manually mock objects instead of just declaring mocks using `@Mock`.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/MiniOzoneClusterImpl.java (hunk 10 above)
Comment (duongkame) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/MiniOzoneClusterImpl.java]: They're not intentional, reverted.

File: hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockInputStream.java (hunk 12 above)
Comment (duongkame) [reply hadoop-hdds/client/src/test/java/org/apache/hadoop/hdds/scm/storage/TestBlockInputStream.java:304]: That test was to confirm the exception handling for IOException, which is now tested in the parameterized `testReadNotRetriedOnOtherException` and `testRefreshOnReadFailure`.

File: hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java (hunk 4 above)
Comment (duongkame) [reply hadoop-ozone/integration-test/src/test/java/org/apache/hadoop/ozone/om/TestOmContainerLocationCache.java:124]: Thanks for the very detailed review, @GeorgeJahad .

**Comment by duongkame** [comment]
> There is a call to looupKey here:
> 
> https://github.com/apache/ozone/blob/1ea4c2adc640a02508696bf346c7f1f7b6e0afd9/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java#L1264
> 
> Should that be changed to getKeyInfo()?

Yes, I cleaned that up as well.


> There is a call to lookupFile() here:
> 
> https://github.com/apache/ozone/blob/1ea4c2adc640a02508696bf346c7f1f7b6e0afd9/hadoop-ozone/client/src/main/java/org/apache/hadoop/ozone/client/rpc/RpcClient.java#L1683
> 
> 
> But lookupFile() doesn't appear to use the cache. Should it?


Thanks, `lookupFile` will be another path we will tackle for OFS. I think it's to contain the scope of this PR for key reads. 
The file read path can probably be deprecated and merged to key read, but we'll see.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java [hunk 13]
```diff
@@ -46,6 +46,7 @@
 import org.apache.hadoop.security.token.Token;
 
 import com.google.common.annotations.VisibleForTesting;
+import org.apache.ratis.thirdparty.io.grpc.Status;
```
Comment (kerneltime) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java:49]: We need to establish is this is the right import.

Comment (adoroszlai) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java:49]: It is, if you want to catch / process exception from Ratis.

File: hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java [hunk 14]
```diff
@@ -187,19 +192,19 @@ public synchronized void initialize() throws IOException {
...
     if (refreshPipelineFunction != null) {
       LOG.debug("Re-fetching pipeline for block {}", blockID);
-      Pipeline newPipeline = refreshPipelineFunction.apply(blockID);
-      if (newPipeline == null || newPipeline.sameDatanodes(pipeline)) {
-        LOG.warn("No new pipeline for block {}", blockID);
-        throw cause;
-      } else {
-        LOG.debug("New pipeline got for block {}", blockID);
-        this.pipeline = newPipeline;
-      }
```
Comment (adoroszlai) [inline hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java:202]: @duongkame May I ask why the check for null/same pipeline was removed?

Comment (duongkame) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java:202]: @adoroszlai, sorry I missed this. My intention was to remove the check for the same pipeline because it just doesn't make sense and fails some test cases. (Client talks to datanode and get an error, it then recontacts to OM and SCM and SCM say the current pipeline is ok, hence, the previous error with datanode may be just a temporary one, so just retry if the policy still allows).

I wasn't aware of any case where the returned pipeline can be null, think I talked to someone to confirm it as well. Obviously you've caught a problem with it, as per HDDS-7982. Thanks.

Comment (adoroszlai) [reply hadoop-hdds/client/src/main/java/org/apache/hadoop/hdds/scm/storage/BlockInputStream.java:202]: No problem.  The fix for HDDS-7982 restores the null check, but avoids throwing exception for same pipeline case.

On the positive side, this change has also fixed HDDS-7089 where read after datanode restart was failing intermittently.  It turned out that `sameDatanodes` ignores IP address change, only checks UUIDs.
